File: cmd/acs/doctor.go
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"os"
	"text/tabwriter"

	"github.com/tuntran/agentcodeswitch/internal/doctor"
	"github.com/tuntran/agentcodeswitch/internal/profile"
)

// cmdDoctor self-checks the installation.
//
// Exit status is part of the contract so scripts can use it: 0 when everything
// passes, 1 when a check failed, 2 when the config itself could not be read.
func cmdDoctor(args []string) error {
	fs := flag.NewFlagSet("doctor", flag.ContinueOnError)
	deep := fs.Bool("deep", false, "also read credentials (macOS will prompt the first time)")
	asJSON := fs.Bool("json", false, "print JSON")
	if err := parseFlags(fs, args); err != nil {
		return err
	}

	report := doctor.Run(profile.DefaultCredStore(), *deep)

	if *asJSON {
		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")
		if err := enc.Encode(report); err != nil {
			return err
		}
	} else {
		printDoctor(report)
	}

	switch {
	case report.ConfigError != "":
		return exitError{code: 2, err: errors.New("config could not be read")}
	case !report.OK():
		return exitError{code: 1, err: errors.New("doctor found problems")}
	}
	return nil
}

func printDoctor(r doctor.Report) {
	if r.ConfigError != "" {
		fmt.Fprintln(os.Stderr, "config: "+r.ConfigError)
		return
	}
	if len(r.Profiles) == 0 {
		fmt.Println("no profiles yet -- run `acs add <name>`")
		return
	}

	tw := tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
	for _, p := range r.Profiles {
		for i, c := range p.Checks {
			name := p.Name
			if i > 0 {
				name = "" // group visually under the profile
			}
			_, _ = fmt.Fprintf(tw, "%s\t%s %s\t%s\n", name, mark(c.Status), c.Name, c.Detail)
		}
	}
	_ = tw.Flush()

	// Actions print after the table so they are not squeezed into a column.
	// Every failure has to answer "what do I do now?" -- a report that only says
	// "fail" takes away the ability to fix it.
	for _, p := range r.Profiles {
		for _, c := range p.Checks {
			if c.Status != doctor.StatusOK && c.Action != "" {
				fmt.Printf("\n%s %s: %s\n  -> %s\n", mark(c.Status), p.Name, c.Detail, c.Action)
			}
		}
	}

	if len(r.Orphans) > 0 {
		// Informational: another tool that sets CLAUDE_CONFIG_DIR produces items
		// that look identical to a drifted acs literal, and acs cannot tell them
		// apart. Say so rather than implying something is broken.
		fmt.Printf("\nkeychain items no acs profile claims: %d\n", len(r.Orphans))
		for _, o := range r.Orphans {
			fmt.Println("  " + o)
		}
		fmt.Println("  These belong to another tool, or to a profile whose config dir literal changed." +
			"\n  If a profile above shows a literal or keychain failure, this is the credential it lost." +
			"\n  Otherwise leave them alone.")
	}
	if !r.Deep {
		fmt.Println("\nrun `acs doctor --deep` to check scopes and expiry (reads secrets)")
	}
}

func mark(s doctor.Status) string {
	switch s {
	case doctor.StatusOK:
		return "✓"
	case doctor.StatusWarn:
		return "!"
	default:
		return "✗"
	}
}

File: frontend/src/views/doctor.ts
import { runDoctor } from "../bindings";
import type { CheckView, DoctorView } from "../types";

// Doctor exists because the failure modes it catches are silent and point the wrong
// way: a drifted config dir literal presents as "I got randomly logged out". So every
// failing check here has to name the fix, not just the fault.

export function renderDoctorView(container: HTMLElement): void {
  const header = document.createElement("header");
  header.className = "page-header";
  const title = document.createElement("h1");
  title.textContent = "Doctor";
  header.append(title);

  const body = document.createElement("div");
  body.className = "stack";
  container.replaceChildren(header, body);

  const shallow = document.createElement("button");
  shallow.type = "button";
  shallow.className = "btn btn--ghost";
  shallow.textContent = "Run checks";
  shallow.addEventListener("click", () => void run(body, false));

  // --deep reads the credential itself, so macOS shows a Keychain prompt. Saying so
  // in advance is the difference between an expected dialog and an alarming one.
  const deep = document.createElement("button");
  deep.type = "button";
  deep.className = "btn btn--ghost";
  deep.textContent = "Run deep checks (macOS will ask for Keychain access)";
  deep.addEventListener("click", () => void run(body, true));

  header.append(shallow, deep);
  void run(body, false);
}

async function run(body: HTMLElement, deep: boolean): Promise<void> {
  body.replaceChildren(note(deep ? "Reading credentials…" : "Checking…"));
  let report: DoctorView;
  try {
    report = await runDoctor(deep);
  } catch (err) {
    body.replaceChildren(banner(String(err), "fault"));
    return;
  }
  body.replaceChildren(...renderReport(report));
}

function renderReport(report: DoctorView): HTMLElement[] {
  if (report.configError) {
    return [
      banner(`The acs config could not be read: ${report.configError}`, "fault"),
    ];
  }
  if (report.profiles.length === 0) {
    return [note("No profiles to check yet.")];
  }

  const out: HTMLElement[] = [
    banner(
      report.ok
        ? "Every check passed."
        : "Some checks failed. Each one below says what to do about it.",
      report.ok ? "info" : "attention",
    ),
  ];

  for (const profile of report.profiles) {
    const card = document.createElement("article");
    card.className = "card";
    card.dataset.profile = profile.name;

    const title = document.createElement("h3");
    title.textContent = profile.name;
    card.append(title);

    for (const check of profile.checks) {
      card.append(checkRow(check));
    }
    out.push(card);
  }

  if (report.orphans.length > 0) {
    // Informational, and worded that way on purpose. Any other tool that sets
    // CLAUDE_CONFIG_DIR leaves a credential acs cannot distinguish from a drifted
    // literal, so telling the user to delete it would mean telling them to break a
    // working install of something else. Matches the CLI and doctor.go's stance.
    const orphans = document.createElement("div");
    orphans.className = "banner banner--info";
    const text = document.createElement("p");
    text.className = "banner__text";
    text.textContent =
      `${report.orphans.length} stored Claude Code credential(s) no acs profile claims: ` +
      `${report.orphans.join(", ")}. These belong to another tool, or to a profile whose ` +
      `config directory literal changed. If a profile above shows a literal or keychain ` +
      `failure, this is the credential it lost. Otherwise leave them alone.`;
    orphans.append(text);
    out.push(orphans);
  }

  if (!report.deep) {
    out.push(
      note(
        "Scopes and token expiry are only checked by the deep run, which reads secrets.",
      ),
    );
  }
  return out;
}

function checkRow(check: CheckView): HTMLElement {
  const row = document.createElement("div");
  row.className = `check check--${check.status}`;
  row.dataset.check = check.name;

  const mark = document.createElement("span");
  mark.className = "check__mark";
  mark.textContent = check.status === "ok" ? "✓" : check.status === "warn" ? "!" : "✗";
  row.append(mark);

  const name = document.createElement("span");
  name.className = "check__name";
  name.textContent = check.name;
  row.append(name);

  const detail = document.createElement("span");
  detail.className = "check__detail";
  detail.textContent = check.detail;
  row.append(detail);

  if (check.status !== "ok" && check.action) {
    const action = document.createElement("p");
    action.className = "check__action";
    action.textContent = `→ ${check.action}`;
    row.append(action);
  }
  return row;
}

function banner(text: string, variant: "info" | "attention" | "fault"): HTMLElement {
  const el = document.createElement("div");
  el.className = `banner banner--${variant}`;
  const p = document.createElement("p");
  p.className = "banner__text";
  p.textContent = text;
  el.append(p);
  return el;
}

function note(text: string): HTMLElement {
  const p = document.createElement("p");
  p.className = "note";
  p.textContent = text;
  return p;
}
